client/src/main/scala/com.crobox.clickhouse/stream/ClickhouseSink.scala

```diff
@@ -31,22 +44,50 @@ object ClickhouseSink extends LazyLogging {
       .getOrElse(indexerGeneralConfig)
     val batchSize     = mergedIndexerConfig.getInt("batch-size")
     val flushInterval = mergedIndexerConfig.getDuration("flush-interval").getSeconds.seconds
-    Flow[Insert]
+    Flow[TableOperation]
       .groupBy(Int.MaxValue, _.table)
       .groupedWithin(batchSize, flushInterval)
-      .mapAsyncUnordered(mergedIndexerConfig.getInt("concurrent-requests"))(inserts => {
-        val table       = inserts.head.table
-        val insertQuery = s"INSERT INTO $table FORMAT JSONEachRow"
-        val payload     = inserts.map(_.jsonRow)
-        logger.debug(s"Inserting ${inserts.size} entries in table: $table. Group Within: ($batchSize - $flushInterval)")
-        client
-          .execute(insertQuery, payload.mkString("\n"))
-          .recover {
-            case ex => throw ClickhouseIndexingException("failed to index", ex, payload, table)
-          }
-          .map(_ => inserts)
+      .mapAsync(mergedIndexerConfig.getInt("concurrent-requests"))(operations => {
+        val table = operations.head.table
+        logger.debug(
+          s"Executing ${operations.size} operations on table: $table. Group Within: ($batchSize - $flushInterval)"
+        )
+
+        insertOp(operations.collect { case o: Insert                     => o }, client)
```

Comment:
Currently, you're traversing the same collection/batch of operations *twice*. This is costly and should be rewritten.
Please google for solutions, if not found, rely on two mutable array buffers (created outside the method and which are cleared in each batch). The big advancement of having these buffers only created once and cleared in each batch is memory consumption

Reply:
I changed the code to use ArrayBuffer as suggested. Another solution that can be also used is using the foldLeft to split the batch into two lists like, it will be something like this:

```
        val splitOps = operations.foldLeft((List[Insert](), List[Optimize]()))(
          (x, y) =>
            y match {
              case o: Insert   => (x._1 ++ Seq(o), x._2)
              case o: Optimize => (x._1, x._2 ++ Seq(o))
              case _           => (x._1, x._2)
          }
        )

        insertOp(splitOps._1, client)
          .flatMap(_ => optimizeOp(splitOps._2, client))
          .map(_ => operations)
```

Reply:
Yes, this is indeed a nice solution!